### ISSUE TITLE
ZFS root pool with Systemd Dracut initram/initrd fails to load pools

### ISSUE THREAD

**Issue by rmackinnon** [openzfs/zfs#12329]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| Gentoo
Distribution Version	| 17.1
Linux Kernel	| 5.4.108
Architecture	| x86_64
ZFS Version	| zfs-2.0.5-r0-gentoo
SPL Version	| spl-2.0.5-r0-gentoo

### Describe the problem you're observing
After building a new initramfs, and booting with it, the system drops into an emergency mode with no pools imported.

### Diagnosed root cause
Diagnosis showed several stages caused this issue.
1. Dracut triggers the module `90zfs` durring image creation of the initramfs, where the file `/etc/zfs/zpool.cache` is copied into the initramfs. On most, if not all, systems this file is a non-zero size, and is key to the root of the issue.
2. On some systems Dracut then softlinks `zfs-import.service` to `zfs-import-cache.service`.
3. The system reboots into the new initramfs, where `zfs-import.target` calls `zfs-import-cache.service`, which finds  `/etc/zfs/zpool.cache` and fails to load any pools, and immediately bails out to an emergency console as it is required to reach the target. Even if  Systemd `zfs-import.target.wants` continues and attempts to start `zfs-import-scan.service`, the service fails when it finds the non-zero file `/etc/zfs/zpool.cache`.
4. Once in a manual emergency mode, zeroing `/etc/zfs/zpool.cache`, and restarting `zfs-import-scan.service` results in proper import of pools, and allows for `zfs-mount.service` and `sysroot.mount` to start.
5. Issuing `systemctl start initrd-switch-root.target` would allow for a user to continue boot into their OS at this point.

In some instances after the prior, some sub-volumes fail to mount even if they are marked to auto-mount.

### Proposed Fix
Alteration of the Dracut `module-setup.sh` to install the following:
/lib/systemd/system/zfs.target -> ${initdir}/$systemdsystemunitdir/initrd-switch-root.target.wants/
/lib/systemd/system/zfs-mount.service -> ${initdir}/$systemdsystemunitdir/initrd-switch-root.target.wants/
/lib/systemd/system/zfs.target -> ${initdir}/$systemdsystemunitdir/local-fs.target.wants/

### Potentially related issues:
#9621 #11249 #10941

**Comment by behlendorf** [comment]
You might try the recent and 2.1.0 tag, I know some improvements were made in this area but I'm not sure if this specific issue was resolved.

**Comment by rmackinnon** [comment]
@behlendorf, thanks for the reply.  I'll try labing it up in a VM and see if it helps and report back here with my findings. The machine I found that issue on has been commissioned into service, so I won't be pushing to 2.1.0 on it till I've tested anyways.

My work-around ended up rewriting the `module-setup.sh` to add and manage the files to be added into the initramfs. There is an (unrelated?) issue for auto mounting volumes after the system transitions to the OS, but I'll open a new issue for that.
